nucliadb_node/src/shards/shard_reader.rs

```diff
@@ -397,108 +398,62 @@ impl ShardReader {
-
-        let relation_request = RelationSearchRequest {
-            shard_id: search_request.shard.clone(),
-            prefix: search_request.relation_prefix.clone(),
-            subgraph: search_request.relation_subgraph,
-            ..Default::default()
-        };
-        let relation_reader_service = self.relation_reader.clone();
-        let relation_task = move || Some(relation_reader_service.search(&relation_request));
+        // Apply pre-filtering to the query plan
+        self.text_reader.apply_pre_filter(&mut query_plan)?;
 
-        let info = info_span!(parent: &span, "text search");
-        let text_task = || run_with_telemetry(info, text_task);
-        let info = info_span!(parent: &span, "paragraph search");
-        let paragraph_task = || run_with_telemetry(info, paragraph_task);
-        let info = info_span!(parent: &span, "vector search");
-        let vector_task = || run_with_telemetry(info, vector_task);
-        let info = info_span!(parent: &span, "relations search");
-        let relation_task = || run_with_telemetry(info, relation_task);
+        // Run the rest of the plan
+        let text_task = query_plan.texts_request.map(|mut request| {
+            request.id = search_id.clone();
+            let text_reader_service = self.text_reader.clone();
+            let info = info_span!(parent: &span, "text search");
+            let task = move || text_reader_service.search(&request);
+            || run_with_telemetry(info, task)
+        });
+
+        let paragraph_task = query_plan.paragraphs_request.map(|mut request| {
+            request.id = search_id.clone();
+            let paragraph_reader_service = self.paragraph_reader.clone();
+            let info = info_span!(parent: &span, "paragraph search");
+            let task = move || paragraph_reader_service.search(&request);
+            || run_with_telemetry(info, task)
+        });
+
+        let vector_task = query_plan.vectors_request.map(|mut request| {
+            request.id = search_id.clone();
+            let vector_reader_service = self.vector_reader.clone();
+            let info = info_span!(parent: &span, "vector search");
+            let task = move || vector_reader_service.search(&request);
+            || run_with_telemetry(info, task)
+        });
+
+        let relation_task = query_plan.relations_request.map(|request| {
+            let relation_reader_service = self.relation_reader.clone();
+            let info = info_span!(parent: &span, "relations search");
+            let task = move || relation_reader_service.search(&request);
+            || run_with_telemetry(info, task)
+        });
 
         let mut rtext = None;
         let mut rparagraph = None;
         let mut rvector = None;
         let mut rrelation = None;
 
         crossbeam_thread::scope(|s| {
-            if !skip_fields {
-                s.spawn(|_| rtext = text_task());
+            if let Some(task) = text_task {
```

Comment:
we should keep the skip conditions?

Reply:
That logic is part of the query planner now

Reply:
ouch, right